### PR TITLE
Performance optimizations for NIST p256 curve

### PR DESCRIPTION
**Pull request by krizhanovsky**
* Montgomery arithmetic 
* Replace mixed coordinates addition with addition in Jacobian coordinates for unknown point multiplication.
* Introduce TLS_CONST_TIME compilation option to use constant time EC computations instead of point randomization.
* Optimize memory usage in memory pools and some EC computations, reduce memory copyings.
* Remove secp384 (#1335)
* Workaround for #1404

### CODE REVIEW

**Review by i-rinat — APPROVED** [review]
:see_no_evil:

File: tls/ec_p256.c [hunk 1]
```diff
@@ -1073,10 +1073,13 @@ ecp256_select_comb(Ecp256Point *r, const EcpXY T[], unsigned char t_len,
...
  * For w=4,D=4S+4M,A=8M+3S,S=0.8M this gives about 1126*M, which with the
- * cost of ecp256_precompute_comb() gives ~2130*M + 2*I, which is ~170-270*M
- * better than the number for Jacobian cure shapes with 2*I in
+ * cost of ecp256_precompute_comb() gives ~1392+ 3*I, which is ~900-1000*M
+ * better than the number for Jacobian cure shapes with 3*I in
  * "Analysis and optimization of elliptic-curve single-scalar multiplication",
  * by Bernstein & Lange, 2007.
+ *
+ * Bigger value for w gives us smaller M total ops in the function, but larger
+ * value for ecp256_double_jac_n(). Current w=4 is the minimum total value.
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply tls/ec_p256.c]: Yeah, the most efficient `w` value is still for TODO for upcoming changes: there is TODO comment in `ecp256_mul_comb()`:
```
         * TODO #1064 at the moment we have S = 0.9*M and very expensive I,
         * so recheck the optimal w value after the optimizations.
```

File: tls/t/test_mpi_math.c [hunk 2]
```diff
@@ -560,6 +559,99 @@ mpi_elementary(void)
...
+	r[2] = (t[5] << 32) | t[4];
+	r[3] = (t[7] << 32) | t[6];
+}
+
+static void
+mont_basic(void)
+{
+	const unsigned long a[8] = {1, 0, 2, 3}, b[4] = {4, 5, 6, 7};
+	unsigned long am[8] = {}, bm[4] = {};
+	unsigned long r[8] = {5}; /* only 4 limbs are used, fill with garbage */
```
Comment (ikoveshnikov) [inline tls/t/test_mpi_math.c:629]: This doesn't do what is stated: the `r[]` array is not filled with garbage, first element is filled with 5, others will be zeroed.

File: tls/t/test_rsa.c [hunk 3]
```diff
@@ -83,7 +82,7 @@ rsa_sign(void)
 	TlsRSACtx *rsa;
 	unsigned char hash[PT_LEN], sig[KEY_LEN];
 
-	EXPECT_FALSE(!(mp = ttls_mpi_pool_create(TTLS_MPOOL_ORDER, GFP_KERNEL)));
+	EXPECT_FALSE(!(mp = ttls_mpi_pool_create(0, GFP_KERNEL)));
```
Comment (ikoveshnikov) [inline tls/t/test_rsa.c:85]: We have `EXPECT_OK` and `EXPECT_NOT_NULL/EXPECT_NULL`.

File: tls/t/tgen_ec256.c [hunk 4]
```diff
@@ -29,18 +29,39 @@
...
+
+static void
+tgen256_set_zero(Ecp256Point *p)
+{
+	ecp256_lset(p->x, 1);
+	ecp256_lset(p->y, 1);
+	ecp256_lset(p->z, 0);
+}
+
+
```
Comment (ikoveshnikov) [inline tls/t/tgen_ec256.c:48]: Extra lines?

**Review by ikoveshnikov — APPROVED** [review]

File: tls/ec_p256.c [hunk 5]
```diff
@@ -814,87 +769,115 @@ ecp256_double_jac_n(Ecp256Point *r, const Ecp256Point *p)
...
+	mpi_sub_mod_p256_x86_64(t3, t3, t4);
+
+	/* Y3 = (U1 * H^2 - X3) * R - S1 * H^3 */
+	mpi_sub_mod_p256_x86_64(t2, t2, t3);
+	mpi_mul_mont_mod_p256_x86_64(t2, t2, tr);
+	mpi_mul_mont_mod_p256_x86_64(s1, s1, h);
+	mpi_sub_mod_p256_x86_64(t2, t2, s1);
+
+	ecp256_copy(r->x, t3);
+	ecp256_copy(r->y, t2);
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply tls/ec_p256.c:877]: The cool thing about the update assembly routines is that they don't require double size of the arguments, so we can use much smaller stack memory. It doesn't affect the performance though.

File: tls/ec_p256.c [hunk 6]
```diff
@@ -814,87 +769,115 @@ ecp256_double_jac_n(Ecp256Point *r, const Ecp256Point *p)
...
+ * multiplications, so SIMD can be applied to compute 2 operations in
+ * parallel (see [5] chapter 4.0.1 and references). Another option (e.g.
+ * Algorithm 4 from [5] or "Bipartite Modular Multiplication" by Kaihara) is
+ * to parallelize different phases of single operation computation process
+ * using interleaved Montgomery computation. Also Faz-Hernandez proposed
+ * several approaces in parallel computations of whole doubling and addition
+ * functions.
+ */
+static void
+ecp256_add_jac(Ecp256Point *r, const Ecp256Point *p, const Ecp256Point *q)
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply tls/ec_p256.c:830]: I added several restrict qualifiers, however is some cases we do have calls with the same result and fist arguments like `ecp256_double_jac(r, r)`

File: tls/t/test_rsa.c (hunk 3 above)
Comment (krizhanovsky) [reply tls/t/test_rsa.c:85]: Actually, the crypto unit tests don't use Tempesta FW's unit test framework and mock the macros. Anyway, I added the mock for `EXPECT_NOT_NULL`

**Pull request merged**